optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/AbstractRuleContext.java

```diff
@@ -88,35 +94,108 @@ protected AbstractRuleContext(ViewItem<?>... viewItems) {
-        scoreHolder.impactScore((RuleContext) drools, justifications);
+    protected static void runConsequence(DroolsConstraint<?> constraint, Drools drools,
+            IntImpactExecutor impactExecutor, int impact, JustificationsSupplier justificationsSupplier) {
+        constraint.assertCorrectImpact(impact);
+        AgendaItem<?> agendaItem = (AgendaItem<?>) ((RuleContext) drools).getMatch();
+        UndoScoreImpacter undo = impactExecutor.execute(impact, justificationsSupplier);
+        agendaItem.setCallback(undo);
     }
 
-    protected static void impactScore(DroolsConstraint<?> constraint, Drools drools, AbstractScoreHolder<?> scoreHolder,
-            int impact, Object... justifications) {
+    protected static void runConsequence(DroolsConstraint<?> constraint, Drools drools,
+            LongImpactExecutor impactExecutor, long impact, JustificationsSupplier justificationsSupplier) {
         constraint.assertCorrectImpact(impact);
-        scoreHolder.impactScore((RuleContext) drools, impact, justifications);
+        AgendaItem<?> agendaItem = (AgendaItem<?>) ((RuleContext) drools).getMatch();
+        UndoScoreImpacter undo = impactExecutor.execute(impact, justificationsSupplier);
+        agendaItem.setCallback(undo);
     }
 
-    protected static void impactScore(DroolsConstraint<?> constraint, Drools drools, AbstractScoreHolder<?> scoreHolder,
-            long impact, Object... justifications) {
+    protected static void runConsequence(DroolsConstraint<?> constraint, Drools drools,
+            BigDecimalImpactExecutor impactExecutor, BigDecimal impact, JustificationsSupplier justificationsSupplier) {
         constraint.assertCorrectImpact(impact);
-        scoreHolder.impactScore((RuleContext) drools, impact, justifications);
+        AgendaItem<?> agendaItem = (AgendaItem<?>) ((RuleContext) drools).getMatch();
+        UndoScoreImpacter undo = impactExecutor.execute(impact, justificationsSupplier);
+        agendaItem.setCallback(undo);
     }
 
-    protected static void impactScore(DroolsConstraint<?> constraint, Drools drools, AbstractScoreHolder<?> scoreHolder,
-            BigDecimal impact, Object... justifications) {
-        constraint.assertCorrectImpact(impact);
-        scoreHolder.impactScore((RuleContext) drools, impact, justifications);
+    protected static IntImpactExecutor buildIntImpactExecutor(WeightedScoreImpacter scoreImpacter) {
+        if (scoreImpacter instanceof IntWeightedScoreImpacter) {
+            return ((IntWeightedScoreImpacter) scoreImpacter)::impactScore;
+        } else if (scoreImpacter instanceof LongWeightedScoreImpacter) {
+            return ((LongWeightedScoreImpacter) scoreImpacter)::impactScore;
+        } else if (scoreImpacter instanceof BigDecimalWeightedScoreImpacter) {
+            return (impact, justificationsSupplier) -> ((BigDecimalWeightedScoreImpacter) scoreImpacter)
+                    .impactScore(BigDecimal.valueOf(impact), justificationsSupplier);
```

Comment:
Why do we care about the `Long` and the `BigDecimal` here?

Reply:
This is a mess. To explain, consider the following situations:

    penalize(..., (int) 1)
    penalizeLong(..., (int) 1)
    penalizeLong(..., (long) 1)
    penalizeBigDecimal(..., (int) 1)
    penalizeBigDecimal(..., (long) 1)
    penalizeBigDecimal(..., BigDecimal.ONE)

All of these are possible, and all of these require different handling, so that `int` is never auto-boxed into `Integer` etc.